hotel_tourist_tax/models/hotel_folio.py

```diff
@@ -0,0 +1,75 @@
+# See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models
+
+
+class HotelFolio(models.Model):
+    _inherit = "hotel.folio"
+
+    @api.multi
+    def write(self, vals):
+        res = super(HotelFolio, self).write(vals)
+        for rec in self:
+            if (
+                bool(vals.get("duration"))
+                or bool(vals.get("adults"))
+                or bool(vals.get("children"))
+            ) and not bool(
+                vals.get("service_lines")  # prevent recursion
+                or rec.invoice_status == "invoiced"
+                or bool(rec.hotel_invoice_id)
+            ):
+                rec.update_tourist_tax()
+        return res
+
+    @api.model
+    def create(self, vals):
+        res = super().create(vals)
+        res.update_tourist_tax()
+        return res
+
+    @api.onchange("duration", "adults", "children")
+    def onchange_fieldname(self):
+        self.update_tourist_tax()
+
+    @api.multi
+    def update_tourist_tax(self):
+        self.ensure_one()
+
+        if not self.order_id:
+            return
+
+        # Remove existing taxes
+        for service_line in self.service_lines:
+            if service_line.product_id in self.env["hotel.tourist.tax"].search(
+                []
+            ).mapped("hotel_service_id.product_id"):
```

Comment:
for clarity and performance (sql request are sent at each loop), you could prepare the tax product list outside the loop.